core/arch/arm/kernel/early_ta.c

```diff
@@ -0,0 +1,244 @@
+		/* TA is compressed */
+		st = decompression_init(&handle->strm, ta);
+		if (!st) {
+			free(handle);
+			return TEE_ERROR_BAD_FORMAT;
+		}
+
+	}
+	handle->early_ta = ta;
+	*h = handle;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result early_ta_get_size(const struct user_ta_store_handle *h,
+				    size_t *size)
+{
+	const struct early_ta *ta = h->early_ta;
+
+	if (ta->uncompressed_size)
+		*size = ta->uncompressed_size;
+	else
+		*size = ta->size;
+	return TEE_SUCCESS;
+}
+
+static TEE_Result read_uncompressed(struct user_ta_store_handle *h, void *data,
+				    size_t len)
+{
+	uint8_t *src = (uint8_t *)h->early_ta->ta + h->offs;
+
+	if (h->offs + len > h->early_ta->size)
+		return TEE_ERROR_BAD_PARAMETERS;
+	if (data)
+		memcpy(data, src, len);
+	h->offs += len;
+	return TEE_SUCCESS;
+}
+
+static TEE_Result read_compressed(struct user_ta_store_handle *h, void *data,
+				  size_t len)
+{
+	z_stream *strm = &h->strm;
+	size_t total = 0;
+	uint8_t *tmpbuf = NULL;
+	TEE_Result ret;
+	size_t out;
+	int st;
+
+	if (data) {
+		strm->next_out = data;
+		strm->avail_out = len;
+	} else {
+		/*
+		 * inflate() does not support a NULL strm->next_out. So, to
+		 * discard data, we have to allocate a temporary buffer. 1K
+		 * seems reasonable.
+		 */
+		strm->avail_out = MIN(len, 1024U);
+		tmpbuf = malloc(strm->avail_out);
+		if (!tmpbuf) {
+			DMSG("Could not allocate temporary buffer");
+			return TEE_ERROR_OUT_OF_MEMORY;
+		}
+		strm->next_out = tmpbuf;
+	}
+	/*
+	 * Loop until we get as many bytes as requested, or an error occurs.
+	 * inflate() returns:
+	 * - Z_OK when progress was made, but neither the end of the input
+	 *   stream nor the end of the output buffer were met
+	 * - Z_BUF_ERROR when there is still input to process but the output
+	 *   buffer is full
+	 * - Z_STREAM_END when the end of the intput stream was reached
+	 */
+	do {
+		out = strm->total_out;
+		st = inflate(strm, Z_SYNC_FLUSH);
+		out = strm->total_out - out;
+		total += out;
+		FMSG("%zu bytes", out);
+		if (!data) {
+			strm->next_out = tmpbuf;
+			strm->avail_out = MIN(len - total, 1024U);
+		}
+	} while ((st == Z_OK || st == Z_BUF_ERROR) && (total != len));
+	if (st != Z_OK && st != Z_STREAM_END && st != Z_BUF_ERROR) {
+		DMSG("TA decompression error (%d)", st);
+		ret = TEE_ERROR_GENERIC;
+		goto out;
+	}
+	ret = TEE_SUCCESS;
+out:
+	if (!data)
+		free(tmpbuf);
+	return ret;
+}
+
+static TEE_Result early_ta_read(struct user_ta_store_handle *h, void *data,
+				size_t len)
+{
+	if (h->early_ta->uncompressed_size)
+		return read_compressed(h, data, len);
+	else
+		return read_uncompressed(h, data, len);
+}
+
+static void early_ta_close(struct user_ta_store_handle *h)
+{
+	if (h->early_ta->uncompressed_size)
+		inflateEnd(&h->strm);
+	free(h);
+}
+
+static struct user_ta_store_ops ops = {
+	.description = "early TA",
+	.open = early_ta_open,
+	.get_size = early_ta_get_size,
+	.read = early_ta_read,
+	.close = early_ta_close,
+	.priority = 5,
```

Comment:
Suggest to define those magic priority values in user_ta.h (i.e `EARLY_TA_STORE_PRIORITY` and `REE_TA_STORE_PRIORITY`).